docs/internals.md

```diff
@@ -1,10 +1,433 @@
+  discard
+
+method view(state: AppState): Widget =
+  gui:
+    Window:
+      CustomBox():
+        Label(text = "I was passed in from the outside") # Uses "add"-adder implicitly by default 
+        Label(text = "Me too!") {.add.} # Uses "add"-adder explicitly
+        Label(text = "Me three!") {.add2.} # Uses "add2"-adder explicitly
+
+when isMainModule:
+  brew(gui(App()))
+```
+
+If no adder is specified, `Widget`s will always be added using the `add` adder. Otherwise the adder defined by the pragma annotation will be used.
+
+### **Hooks**
+
+Hooks are a concept introduced by owlkettle that allows you to execute code throughout a widget's lifecycle, or when an action on one of its fields occurs.
+
+Most hooks are defined only for Widgets, some are defined for both and `property` is only available as a hook for fields.
+
+The available hooks are:
+  - (W) beforeBuild     : Executed only once before the `WidgetState` is created.
+  - (WF) build           : Executed once after `WidgetState` is instantiated from `Widget`. Default values have not yet been applied and will overwrite any values set within this hook.
+  - (W) afterBuild      : Executed only once after the `WidgetState` was created. Is executed *after* all default values have been applied.
+  - (W) connectEvents   : Only relevant for renderables. Executed every time a callback is supposed to be attached to the underlying GTK widget. It defines how to do so.
+  - (W) disconnectEvents: Only relevant for renderables. Executed every time a callback is supposed to be removed from the underlying GTK widget. It defines how to do so.
+  - (WF) update         : Executed every time `WidgetState` is updated by `Widget`.
+  - (F) property        : Executed every time the hook-field changed its value in both the update and build phases.
+  - (F) read            : Used in `Dialog`. Executes every time the state of the underlying GTK-Widget changes.
+
+W: Can act as hook for Widgets
+F: Can act as hook for fields
+
+All hooks have implicit access to a variable called `state`, which contains the `WidgetState`-instance of your particular widget.
+
+With the exception of `read`, all hooks also have implicit access to a variable called `widget`, which is the `Widget` instance.
+
+Generally the `build` and `update` hook are likely to have the highest utility for you. Consult their individual sections for more information.
+
+#### Build Hook
+The `build` hook runs just before any values are assigned to the `WidgetState`.
+
+The intended usecase for build-hooks is adding logic that sets fields on `WidgetState` that don't have default-values.
+
+Example: A Widget may need to load data from elsewhere, via a file or HTTP request. Doing this in the `view` method would rerun the code every time the widget re-renders. But in the build-hook, it only runs once during widget-creation. 
+
+Here a simple code-example:
+
+```nim
+import std/json
+
+type Config = object
+  name: string
+
+viewable MyViewable:
```

Comment:
Maybe make this the `App` to simplify the example.

Reply:
Done. Also added the `json` file so that the example is more complete.